controller/volume_controller.go

```diff
@@ -1091,21 +1103,16 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
 		e.Spec.SalvageRequested = false
 	}
 
-	allFaulted := true
-	for _, r := range rs {
-		if r.Spec.FailedAt == "" {
-			allFaulted = false
-		}
-	}
-	if allFaulted {
+	isAutoSalvageNeeded := vc.getHealthyReplicaCount(rs) == 0 && vc.getFailedReplicaCount(rs) > 0
+	if isAutoSalvageNeeded {
```

Comment:
I like the simplification but are these equivalent?
What's the state on a newly created volume/replica/engine?

Reply:
I think yes, they cover the old logic. When the volume has no healthy replica, there could be 2 cases:
* All replicas have never ever become healthy (both failedAt and healthyAt are empty), e.g. newly created volume
* There are some replicas which were once healthy but now failed -> do auto salvage 

cc @shuo-wu What do you think about this change?

Reply:
This logic refactoring is nice.